pkg/authenticator/request/headerrequest/requestheader.go

```diff
@@ -49,7 +51,12 @@ func (a *Authenticator) AuthenticateRequest(req *http.Request) (*authenticator.R
 		identity.Extra[authapi.IdentityPreferredUsernameKey] = preferredUsername
 	}
 
-	return identitymapper.ResponseFor(a.mapper, identity)
+	res, ok, err := identitymapper.ResponseFor(a.mapper, identity)
+	if res != nil && res.User != nil {
+		audit.AddUsernameAnnotation(req, res.User.GetName())
```

Comment:
I am adding the decisions at the very last moment. You can't change a decision. So I am handling it in the middleware in auth.go

Reply:
All Authenticator.Request functions should get their decision here: https://github.com/openshift/oauth-server/blob/master/pkg/audit/authreq.go#L28-L35=
I ran into cases, where I had a deny, instead of an allow, when adding decisions down in the username logic. As mentioned, it somehow happened that the basic request decision was set to deny and you couldn't change it back to allow, once it reached requestheader.

Reply:
I added a logic to `audit.AddUsernameAnnotation` to only log non-empty strings. So we can be more eager in trying to set the username.

Reply:
This looks like a design flaw. Maybe we could add the name of the provider into the decision annotation? Something like `"authentication.openshift.io/decision.<providerName>"`? Not sure what would be the simplest way to make it easily consumable by, say, `jq`.

Seeing denies/allows of all the attempted IdPs in the event seems to make a lot of sense in my head, I wonder whether you would agree.

Reply:
I think in combination with the OIDC IdP, where getName returns an ID, it would make sense to do so. Let me sync with @pierreprinetti on that.

Reply:
The design of auth-6 is kinda problematic as there are several factors that come into play:
- you can set an annotation only once,
- you have a list of IDPs, so you can't set the decision of the very first IDP,
- the logic is spread around and confusing to follow and
- on error cases we don't return the identity, just an error or false.

The best idea that arose in a chat between us, was to add a username to a custom error and get rid of the bool flag. But this could be a bigger refactoring, as we would violate the authenticator.Request interface.

Reply:
As of our discussion, I assume there is nothing to do here.